baselines/adamw/pytorch/submission.py

```diff
@@ -0,0 +1,164 @@
+    cosine_steps = max(step_hint - hyperparameters.warmup_steps, 1)
+    cosine_decay = CosineAnnealingLR(optimizer, T_max=cosine_steps)
+    return SequentialLR(
+        optimizer,
+        schedulers=[warmup, cosine_decay],
+        milestones=[hyperparameters.warmup_steps])
+
+  optimizer_state['scheduler'] = pytorch_cosine_warmup(
+      workload.step_hint, hyperparameters, optimizer_state['optimizer'])
+
+  return optimizer_state
+
+
+def update_params(workload: spec.Workload,
+                  current_param_container: spec.ParameterContainer,
+                  current_params_types: spec.ParameterTypeTree,
+                  model_state: spec.ModelAuxiliaryState,
+                  hyperparameters: spec.Hyperparameters,
+                  batch: Dict[str, spec.Tensor],
+                  loss_type: spec.LossType,
+                  optimizer_state: spec.OptimizerState,
+                  eval_results: List[Tuple[int, float]],
+                  global_step: int,
+                  rng: spec.RandomState) -> spec.UpdateReturn:
+  """Return (updated_optimizer_state, updated_params, updated_model_state)."""
+  del current_params_types
+  del loss_type
+  del eval_results
+
+  current_model = current_param_container
+  current_model.train()
+  optimizer_state['optimizer'].zero_grad()
+
+  logits_batch, new_model_state = workload.model_fn(
+      params=current_model,
+      augmented_and_preprocessed_input_batch=batch,
+      model_state=model_state,
+      mode=spec.ForwardPassMode.TRAIN,
+      rng=rng,
+      update_batch_norm=True)
+
+  label_smoothing = (
+      hyperparameters.label_smoothing if hasattr(hyperparameters,
+                                                 'label_smoothing') else 0.0)
+  if hasattr(hyperparameters, 'grad_clip'):
+    grad_clip = hyperparameters.grad_clip
+  else:
+    grad_clip = None
+  loss, _ = workload.loss_fn(
+      label_batch=batch['targets'],
+      logits_batch=logits_batch,
+      mask_batch=batch.get('weights'),
+      label_smoothing=label_smoothing)
+
+  loss.backward()
+
+  with torch.no_grad():
+    parameters = [p for p in current_model.parameters() if p.grad is not None]
+    grad_norm = torch.norm(
+        torch.stack([torch.norm(p.grad.detach(), 2) for p in parameters]), 2)
+
+  if grad_clip is not None:
+    torch.nn.utils.clip_grad_norm_(
+        current_model.parameters(), max_norm=grad_clip)
+  optimizer_state['optimizer'].step()
+  optimizer_state['scheduler'].step()
+
+  # Log training metrics - loss, grad_norm, batch_size.
+  if global_step <= 100 or global_step % 500 == 0:
+    if workload.metrics_logger is not None:
+      workload.metrics_logger.append_scalar_metrics(
+          {
+              'loss': loss.item(),
+              'grad_norm': grad_norm.item(),
+          }, global_step)
+    logging.info('%d) loss = %0.3f, grad_norm = %0.3f',
+                 global_step,
+                 loss.item(),
+                 grad_norm.item())
+
```

Comment:
Do you think it makes sense to add `optimizer_state[‘optimizer’].zero_grad(set_to_none=True)` to the line before the `return` in all the pytorch baselines? That should more aggressively free the memory being used by grads. I'm no pytorch pro though so I'll leave it up to you

Reply:
Cool idea, I haven't really seen this anywhere together with setting the gradients to 0 before the backward pass. This could in some cases maybe lead to a small slow-down because memory has to be allocated again before each update, but apparently [not always](https://discuss.pytorch.org/t/freeing-gradients-memory-after-optimizer-step/116346/4). Do we actually care about saving memory in the case of the baselines? They run with the fixed batch sizes anyway.

Reply:
yeah that makes sense, I was playing around with it for criteo but that was it. fine with leaving it out for now.